### ISSUE TITLE
Seq.empty renders as `"EmptyEnumerable"`

### ISSUE THREAD

**Issue by bartelink** [dotnet/fsharp#17864]
`Seq.empty<MyType>` renders as the string `"EmptyEnumerable"` when fed through AspNetCore 8 `ObjectResult`

The same does not happen for `Seq.choose` and other such functions, as they use a different internal impl type

**Expected behavior**

render as `[]`

**Actual behavior**

renders as `"EmptyEnumerable"`

**Known workarounds**

use `Array.empty<MyType>` instead

**Related information**

- happen to be using JSON.NET and a random mix of middlewares and don't have time to root cause deeper, but it surprised me
- `MyType` has a JsonConverter registered
- Did not yet try with `System.Linq.Enumerable.Empty<MyType>()`

**Comment by bartelink** [comment]
I realize this is horrendously incomplete, but I didn't find the condition to be easy to search for, and saying nothing is not ideal either.

If it was me fixing as a quick win without root causing / understanding the required interfaces/moving parts more deeply, I'd make `Seq.empty` delegate to `Array.empty` (which is my actual workaround for now), but obviously there's plenty to be won from an optimal implementation too.

Here's hoping I eventually get to the point of circling back with a trimmed down repro that identifies whether this is a JSON.NET problem, a JsonConverter+JSON.NET problem, or only solvable by implementing a relevant interface on `EmptyEnumerable`

**Comment by Martin521** [comment]
Interesting:

```
> let es: seq<int> = [];;
val es: int seq = []

> let es2: seq<int> = Seq.empty;;
val es2: int seq

> es;;
val it: int seq = []

> es2;;
val it: int seq = seq []
```

**Comment by bartelink** [comment]
I'm not ruling anything in or out
In my context, `[]` is much more 'correct' than a sequence mapping to a `string` JSON value
The specific reasons why fsi will render in a given manner are definitely something that's interesting and relevant, but for me isn't an open and shut answer either.

    > printfn $"{System.Linq.Enumerable.Empty<unit>()}"
    System.Linq.EmptyPartition`1[Microsoft.FSharp.Core.Unit]

(But I have yet to test how that renders either - All I know for a fact is that `Array.empty` renders equivalent to the wrapper that `Seq.choose` yields)

**Comment by Martin521** [comment]
F# type checking infers `es` in the above example to be of type `int list` and `es2` to be of type `ÈmptyEnumerable`.
I don't think there is anything wrong with it.
I guess that in the case that you mention above, using `Seq.choose`, there is a reason why type checking infers a list or array.

**Comment by bartelink** [comment]
Yeah, it's not entirely/only about the static type from the perspective of the compiler though - ultimately `ObjectResult` is not strongly typed (the data is an `obj`), so it's about the interfaces and other aspects of the wrapper type; that's what dictates how JSON.NET and/or AspNetCore's wiring ends up rendering the result. (Seq.empty casts the internalt wrapper type to `seq<'t>`, but that turns out not to be enough)

**Comment by Lanayx** [comment]
I've tried it with `TypedResults.Ok(Seq.empty)` and it fails with the exception 
```
System.NotSupportedException: F# discriminated union serialization is not supported. Consider authoring a custom converter for the type. The unsupported member type is located on type 'Microsoft.FSharp.Collections.IEnumerator+EmptyEnumerable`1[System.Object]'.
```

**Comment by T-Gro** [comment]
For F# DUs which also implement IEnumrable, S.T.J could use IEnumerableDefaultConverter perharps?

Right now, F# DUs are not supported, no matter what the interface.
(even though Lists and Options are also DUs, but they are being special cased by the runtime type detection)

https://github.com/dotnet/runtime/blob/aba819956355c47a9dddbcba8a1aca591f176166/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/FSharp/FSharpTypeConverterFactory.cs#L69

**Comment by vzarytovskii** [comment]
Currently, `Seq.empty` is rendering "correctly" as `EmptyEnumerable` from the perspective of STJ (it chooses to take a type name as its representation), since it's a single-case DU:

```fsharp
// seqcore.fs:

[<NoEquality; NoComparison>]
    type EmptyEnumerable<'T> =

        | EmptyEnumerable

        interface IEnumerable<'T> with
            member _.GetEnumerator() = Empty<'T>()

        interface IEnumerable with
            member _.GetEnumerator() = (Empty<'T>() :> IEnumerator)

// seq.fs:

let empty<'T> = (EmptyEnumerable :> seq<'T>)
```

**Comment by bartelink** [comment]
I get that. But for me STJ and its specfic handling is a red herring here

The point is more that you don't expect as an ASP.NET user to get different results for `[||]`/`[]`/`Seq.empty`/`Array.empty`.

This manifested for me as a very strange bug where I had a Controller body that was doing a Seq.choose and working, but the special case returning Seq.empty acted differently

My hackaround was to use Array.empty. List.empty would have worked too. The reason I logged this is that I'd be surprised if such coincidences would happen if Seq.empty's internal impl e.g. delegated to `Enumerable.Empty`.

**Comment by vzarytovskii** [comment]
> The reason I logged this is that I'd be surprised if such coincidences would happen if Seq.empty's internal impl e.g. delegated to `Enumerable.Empty`.

Yeah, it probably gets confused as it's not get upcasted to `IEnumerable<_>`. I wondering what would happen if you pass it as `Seq.empty :> IEnumerable<_>`?

> As in making it return a different impl type, or giving it different semantics? I accept that one shouldn't just change the impl for a laugh given the subtle effects that any such change in such a core type would have. But by the same token, I bet Enumerable.Empty has had internal change and churn e.g. as part of perf tuning over time.

Mostly first. Different return type will make it backwards incompatible (mostly talking about binary compatibility).

**Comment by bartelink** [comment]
> Yeah, it probably gets confused as it's not get upcasted to IEnumerable<_>. I wondering what would happen if you pass it as Seq.empty :> IEnumerable<_>?

ASP.NET Has an internal ObjectResult that carries it as a 'boxed' `obj`. I'm sure stuff works fine if you let type inference have the result type end up as Task<'T>. 

> Different return type will make it backwards incompatible (mostly talking about binary compatibility).

OK but in this case the static result type is `seq<'t>`, and the proposal is not to change that - it's about the internal type that's yielded and the additional interfaces/characteristics that become apparent via other mechanisms. I believe the impl type in question is internal so there won't be a dependency on that, but it's easy to imagine other random dependencies like expectations of other related interfaces being implemented as you suggest.

**Comment by Martin521** [comment]
Somebody produces an object that has `seq<_>` / `IEnumerable<_>` as runtime type.
If this sequence is empty, than the correct `ToString()` representation is `EmptyEnumerable`.
Why should it render as `[]` or [| |]? It is not a list nor array.
So, the issue is not in how an empty `seq` is rendered, but that somebody somewhere made it a (runtime!) `seq` rather than a `list` or `array`.
(Or, if we are not talking about `ToString()`, but about JSON, it is an issue of the json library.)

**Comment by bartelink** [comment]
You can go make different issues if you like. My problem that made me log this is that:

```
[<HttpGet>
member _.Thing() = task {
    if ... then source |> Seq.choose f
    else Seq.empty
```

Causes a suprising and annoying result, regardless of who did what right or wrong.

And that I suspect that replacing `Seq.empty` with `System.Linq.Enumerable.Empty()` will fix it.

That's an annoying real world usage thing which can be fixed either in ASP.NET or in FSharp.Core. And I don't see a PR that special cases EmptyEnumerable going into the ASP.NET acceptance test suite happening.

I won't consider this issue closed until such time as the above Just Works as it should.

**Comment by T-Gro** [comment]
I really think this has to do with something (S.T.J. most likely, being used by aspnetcore) using a runtime detection.
And when it sees an F# DU at runtime (even though the compile-time type is just seq), it simply calls .ToString().

Whereas if it gets any other seq, which just happens not to be a DU (which should be an implementation detail, but the implementation leaks by runtime inspection via reflection!), it uses a different codepath to render it.


If anyone can confirm that aspnetcore uses S.T.J to produce ObjectResult's string representation, and that it fallbacks to .ToString() when it cannot find a converter - we can prepare a specific fix for it.
(and the fix will help any other F# DU which implements IEnumerable, not just a single one)

**Comment by Martin521** [comment]
> I really think this has to do with something (S.T.J. most likely, being used by aspnetcore) using a runtime detection. And when it sees an F# DU at runtime (even though the compile-time type is just seq), it simply calls .ToString().

In the above example it is not related to a DU, but already happens in `Seq.empty |> ObjectResult`. `ObjectResult` is not generic and stores the empty sequence as `obj`. So, the issue is clearly not in FSharp.Core, but in STJ.

STJ is well known to _not_ well support F# DUs. That's why packages like [FSharp.SystemTextJson](https://www.nuget.org/packages/FSharp.SystemTextJson) exist.

Unfortunately, the discussion [there](https://github.com/dotnet/runtime/issues/55744) is not going well.

**Comment by Lanayx** [comment]
Those are benefits of changing SDU to array in F# Core:

- Fixes related bugs in all current and future serialization libraries
- Improves serialization performance (since serializer can detect array and use .Length instead of trying to enumerate collection)
- Improves new users experience, they won't have "surprising bugs"
- using DU for a collection type doesn't make sense anyway

Downsides:
- Exceptions in code that used private reflection
- Exceptions in code that used binary serializer and stored that somewhere (deprecated since .NET 8)

So in my mind trying to protect the buggy code shouldn't be the goal, while increased stability in performance should be. After all it's Liskov substitution principle - once you expose interface, you are free to change implementation details.

**Comment by vzarytovskii** [comment]
> So in my mind trying to protect the buggy code shouldn't be the goal

You mean one in stj and co? Agree.

**Comment by bartelink** [comment]
NSJ has been serializing JSON results in ASP.NET for a long time
Can we leave STJ out of it please. (And FSharp.STJ is whether or not that fixes matters again is not relevant either)

There are many ways _I_ can fix it; the point is out of the box for Newtonsoft it has a surprising trap.

I'm not saying that anyone broke any law
But I did not get to fall into the pit of success
If the impl happened to be more direct than an interace on a DU, then I would have

I don't care how the problem is fixed, but know that the problem from my perspective is "F# has a funny internal wart that happens to cause a real problem for a real usage". I was seeking to imply "and we should probably remove that unless there's a really good reason for it, or Enumerable.Empty has the same problem". I am not interested in any legal discussion; I fully understand that Seq.empty returns seq<'t> and that it is entitled to do so as it sees fit.

I am also not trying to say that web frameworks that use `obj` internally are the last word. But they are common.

**Comment by T-Gro** [comment]
> "F# has a funny internal wart that happens to cause a real problem for a real usage". 


If you look at the generated IL, it's just a plain empty class implementing the IEnumerable interface. The whole reason this happens is that some libraries are voluntarily choosing to look for `[CompilationMapping(SourceConstructFlags.SumType)]` attribute on the class, which is just metadata. I don't think there is anything else funny about it for a .NET library.


> "and we should probably remove that unless there's a really good reason for it, or Enumerable.Empty has the same problem". 

A change has to justify itself, not the other way around. @Lanayx  mentioned some other arguments - if they can be demonstrated and they indeed bring a notice-able benefit, I would personally be more open to the change. For example a demonstratablybetter perf in common scenarios (examples: plain creation+return, LINQ's .ToArray(), Seq.toArray, Seq.toList, serialization via a chosen tool, iteration).

Those might all be good arguments and will justify the change.
Pleasing a library which is effectively doing runtime detection of an attribute does not make it  as a reason in my opinion.

**Comment by vzarytovskii** [comment]
> For example a demonstratable better perf in common scenarios (examples: plain creation+return, LINQ's .ToArray(), Seq.toArray, Seq.toList, serialization via a chosen tool, iteration).

To be fair, it's hard to imagine improvements can be noticeable, especially at runtime when JIT kicks in. Will be curious too see what are those potential improve as well. Also need to think of there will be any implications in sigdata/inlining when we build versus one fslib, and run with another version. Probably not, but will need verifying.

Regarding "funny internal wart" - this is definitely not what's happening. Codegen is nothing special, just certain libraries serialize things in certain way.

**Comment by bartelink** [comment]
> Regarding "funny internal wart" - this is definitely not what's happening. Codegen is nothing special, just certain libraries serialize things in certain way.

I get that the codegen is nothing special, and that the IEnumerable impl is fine. My only point is that Enumeable.Empty is unlikely to fall prey to this. The 'funny internal wart' is the observed effect - two different impls that dont work out of the box. And the one and only problem is "F# surprisingly OOTB has a wacky thing that happens when you try to use it for normal programming returning an empty sequence".

And the perf has nothing to do with it. I was just suggesting that `Enumerable.Empty` impl has had perf treatments over time even though its ostensibly a very simple thing that doesnt justify messing with. I personally would settle for `Seq.empty` delegating to `Array.empty` and calling it a day myself! (Mainly kidding - I know there are plenty reasons why that wouldn't Just Work)

**Comment by vzarytovskii** [comment]
> > Regarding "funny internal wart" - this is definitely not what's happening. Codegen is nothing special, just certain libraries serialize things in certain way.
> 
> I get that the codegen is nothing special, and that the IEnumerable impl is fine. My only point is that Enumeable.Empty is unlikely to fall prey to this. The 'funny internal wart' is the observed effect - two different impls that dont work out of the box. And the one and only problem is "F# surprisingly OOTB has a wacky thing that happens when you try to use it for normal programming returning an empty sequence".
> 
> And the perf has nothing to do with it. I was just suggesting that `Enumerable.Empty` impl has had perf treatments over time even though its ostensibly a very simple thing that doesnt justify messing with. I personally would settle for `Seq.empty` delegating to `Array.empty` and calling it a day myself! (Mainly kidding - I know there are plenty reasons why that wouldn't Just Work)


I guess we're going in circles here. I understand that serialisation is unexpected for something out-of-the-box in F#. I understand that either we or the library will have to do something. I'm just trying to say that we should really think good about what we want to do, and think about potential side-effects and implications for existing and old tooling.

**Comment by Lanayx** [comment]
Serialization benchmark using `System.Text.Json.JsonSerializer.Serialize`:
```fsharp
let seqEmpty = Seq.empty<int>
let arrEmpty = Array.empty :> seq<int>

[<Benchmark>]
member this.SeqEmpty() =
    for _ in 1..1000 do
        System.Text.Json.JsonSerializer.Serialize(seqEmpty) |> ignore

[<Benchmark>]
member this.ArrayEmpty() =
    for _ in 1..1000 do
        System.Text.Json.JsonSerializer.Serialize(arrEmpty) |> ignore
```

```
BenchmarkDotNet v0.14.0, Windows 11 (10.0.22631.4460/23H2/2023Update/SunValley3)
AMD Ryzen 5 5600H with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK 9.0.100
  [Host]     : .NET 9.0.0 (9.0.24.52809), X64 RyuJIT AVX2 DEBUG
  DefaultJob : .NET 9.0.0 (9.0.24.52809), X64 RyuJIT AVX2


| Method     | Mean     | Error    | StdDev   | Gen0   | Allocated |
|----------- |---------:|---------:|---------:|-------:|----------:|
| SeqEmpty   | 90.78 us | 1.681 us | 1.651 us | 6.5918 |  54.69 KB |
| ArrayEmpty | 89.08 us | 1.742 us | 1.544 us | 3.7842 |  31.25 KB |
```

**Comment by Lanayx** [comment]
Suprisingly `Seq.toArray` works better for Seq.empty, but I believe it could be fixed for this special case
```fsharp
let seqEmpty = Seq.empty<int>
let arrEmpty = Array.empty :> seq<int>

[<Benchmark>]
member this.SeqEmpty() =
    for _ in 1..1000 do
        seqEmpty |> Seq.toArray |> ignore

[<Benchmark>]
member this.ArrayEmpty() =
    for _ in 1..1000 do
        arrEmpty |> Seq.toArray |> ignore
```

```
BenchmarkDotNet v0.14.0, Windows 11 (10.0.22631.4460/23H2/2023Update/SunValley3)
AMD Ryzen 5 5600H with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK 9.0.100
  [Host]     : .NET 9.0.0 (9.0.24.52809), X64 RyuJIT AVX2 DEBUG
  DefaultJob : .NET 9.0.0 (9.0.24.52809), X64 RyuJIT AVX2


| Method     | Mean      | Error     | StdDev    | Gen0   | Allocated |
|----------- |----------:|----------:|----------:|-------:|----------:|
| SeqEmpty   |  5.944 us | 0.1010 us | 0.0789 us | 2.8687 |  23.44 KB |
| ArrayEmpty | 61.335 us | 0.9962 us | 0.9318 us | 2.8076 |  23.44 KB |

```

**Comment by Lanayx** [comment]
`Seq.length` is better with array as expected
```fsharp
let seqEmpty = Seq.empty<int>
let arrEmpty = Array.empty :> seq<int>

[<Benchmark>]
member this.SeqEmpty() =
    for _ in 1..1000 do
        seqEmpty |> Seq.length |> ignore

[<Benchmark>]
member this.ArrayEmpty() =
    for _ in 1..1000 do
        arrEmpty |> Seq.length |> ignore
```
```
BenchmarkDotNet v0.14.0, Windows 11 (10.0.22631.4460/23H2/2023Update/SunValley3)
AMD Ryzen 5 5600H with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK 9.0.100
  [Host]     : .NET 9.0.0 (9.0.24.52809), X64 RyuJIT AVX2 DEBUG
  DefaultJob : .NET 9.0.0 (9.0.24.52809), X64 RyuJIT AVX2


| Method     | Mean      | Error     | StdDev    | Gen0   | Allocated |
|----------- |----------:|----------:|----------:|-------:|----------:|
| SeqEmpty   | 12.570 us | 0.2466 us | 0.2741 us | 2.8687 |   24000 B |
| ArrayEmpty |  2.718 us | 0.0506 us | 0.0449 us |      - |         - |
```

**Comment by Lanayx** [comment]
So overall, changing current implementation to `Array.empty` gives performance boost everywhere. I can create a PR if this change is approved.

**Comment by bartelink** [comment]
Per comments in [EmptyEnumerable](https://github.com/microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs#L2143-L2150), it may be a good idea to consider delegating to [`Enumerable.Empty`](https://github.com/microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs#L1284-L1286) in order to benefit from current and/or future special casing it has.

**Comment by T-Gro** [comment]
Indeed, that would be my suggestion as well.
As of now, it does indeed return a shared singleton empty array, but it does not have to stay like this forever.

**Comment by Lanayx** [comment]
> As of now, it does indeed return a shared singleton empty array

Decompiled sources say it returns `EmptyPartition`
![Image](https://github.com/user-attachments/assets/36127e07-b3d6-4326-9795-cf7a68d03ecb)
But I don't mind using `Enumerable.Empty` anyway instead of SCDU, although this would require check other FSharp.Core array functions to make sure that they end up using optimized methods (like `EmptyPartition.ToArray` etc)

**Comment by vzarytovskii** [comment]
Additional question whether Fable can rely on types when they transpile. Changing type might be braking for them (hopefully they don't rely on EmptyEnumerable directly though).

**Comment by Lanayx** [comment]
No, this won't break anything for Fable, since it relies on the `IEnumerable<'T>` interface, and uses custom implementation, so internal .NET implementation doesn't affect it in any way:

```fsharp
let x = Seq.empty<int>
```
```js
import { empty } from "fable-library-js/Seq.js";

export const x = empty();
```

**Comment by vzarytovskii** [comment]
Good to know
